### DISCUSSION TITLE
Solar temperature differential setting?

### DISCUSSION THREAD

**Discussion by CodeSlinger69** [tagyoureit/nodejs-poolController#925] [Nixie]
Hey folks, I'm still here, just been away from my house and pool for 6 months. @FlaMike, glad to see you are getting your cannibalized Nixie built and your questions answered!

We are now in solar heating season in AZ, and I noticed that when my pool hits the setpoint, the solar turns off as expected. However, the precision on the temperature probe via REM is such that the temperature will fluctuate by several hundredths of a degree, which triggers the solar to reengage. For exmaple, my setpoint is 83, and REM will report a temp of 82.96.

njsPC is smart enough to detect this short cycling situation and will not restart the solar for 60 seconds, to allow time for valves to rotate to where they need to be. It does this for gas heating as well, BTW, to avoid damaging the furnace. But what happens then is that the solar will turn on every minute or so for 10 seconds, as it appears that njsPC does not have temperature differential logic. I could very well be missing something of course. 

As an example of what I mean: Your thermostat for your home furnace has a setpoint, say 72 degrees, and when that setpoint is reached, it will turn the furnace off. But the thermostat also has a temperature differential (3 degrees or 1 degree are typical, some are also programmable). If your differential is say 3 degrees, then the thermostat will not call for heat until the house temperature drops to 69. This is so the furnace is not engaged when the temperature drops to 71.9.

I have attached a replay which shows this behavior. You can see when the water temperature drops 0.04 degrees, njsPC starts thinking about turning the solar back on, but does wait 60 seconds to avoid short cycling.

So the question: Does njsPC have a temperature differential setting, and if not, should it?

[replay.zip](https://github.com/tagyoureit/nodejs-poolController/files/14972372/replay.zip)

**Comment by rstrouse** [comment]
Double check to ensure you have your start and stop temp deltas set up.

**Reply by johnny2678** [reply]
Do you have an OCP? EasyTouch or Intellitouch?

I have EasyTouch and notice pump rpms changing rapidly not that solar heating is active. It just bounced back and forth between the solar speed and the regular spend in short intervals. Haven't had time to look into it but was thinking it was related to solar settings.

Though I'm still getting stacktraces when trying to change solar deltas in dashpanel.

**Reply by CodeSlinger69** [reply]
Here's my config for start/stop deltas. My understanding is that these only control when solar heating is active relative to the solar temperature being observed and the current water temperature. In other words:

- If solar temp > water temp + start delta, then engage solar heating
- If solar temp < water temp - stop delta, then disengage solar heating

I don't think these two deltas have anything to do with the termperature differential question I have? At least I don't see them used in the code where the logic for short cycling is expressed.

![image](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/add40dd5-4fee-4e28-ae95-ce0168c0511a)

**Reply by CodeSlinger69** [reply]
> Do you have an OCP? EasyTouch or Intellitouch?

Nixie. You can see my build [here](https://github.com/tagyoureit/nodejs-poolController/discussions/801).

**Reply by tagyoureit** [reply]
EDIT:  Why is this simple problem so hard?  What I put below isn't correct.  Need some time to think about it.  :(

ORIGINAL: Pull the changes I just pushed up.  The solar heater will now be on when the pool temp is between the set point - start temp delta and  set point + stop temp delta (and of course the solar temp > body temp).  EG the set point is 83 the solar heater should be on when the pool temps are between 76 and 88 (based on -7 start delta and +5 stop temp delta).

Now, cycling may only happen rapidly if the solar temp dips above/below the body temp.  But that's a different problem.

**Reply by CodeSlinger69** [reply]
Interesting, thank you. I'll do some testing.

The logic that incorporates setpoint, water temp, solar temp, start delta, and stop delta are still a bit opaque to me. For anyone else following along, this is what I takeaway from @tagyoureit's changes above:

Turn solar on if:
- water temp < setpoint + stop delta AND
- water temp > setpoint - start delta AND
- solar temp > water temp

If any of those clauses is false, then solar will not be turned on (or turned off if it is already on).

My current setpoint is 83, start delta 7, stop delta 5. Plugging the numbers, my solar should turn on when:
- water temp < 88 AND water temp > 76

88 is too high of a target for me, so I'll set setpoint to 81, start delta to 5, and stop delta to 2, so it looks like this:
- water temp < 83 AND water temp > 76

This logic has a flaw though, which is that I don't care about water temp when solar heating my pool in the spring. The water temp could be 50, and as long as solar temp > water temp + start delta (in the old logic), then solar would engage. The new logic prevents this from happening unless I change the setpoint to a low number and then increase it periodically as the water warms up.

Regardless, I'll test and give it some more thought.

**Reply by CodeSlinger69** [reply]
> EDIT: Why is this simple problem so hard? What I put below isn't correct. Need some time to think about it. :(

My $0.02. The original logic was fine IMHO. I think the only thing needed is a latch to prevent the solar from reengaging after it hits the setpoint. That latch is the "temperature differential" - once the water temp drops by the amount of that differential, then the latch would be unlatched, and the solar would free to turn back on, assuming all the other clauses are also true.

Like I said in my O.P., this is standard logic in household thermostats that control your furnace.

**Reply by tagyoureit** [reply]
> water temp < 88 AND water temp > 76

You realize this statement will _never_ be true.  :) 

I think I've been thinking about this wrong.  I read the Pentair manual for the ump-teenth time and something finally caught my eye.

> Screen (3/3): Solar temperature differential start up and run settings.
Start: Set the temperature differential to start heating from 3° to 9°. For example, if
“Start” is set to 3°, this ensures that the temperature has to deviate by 3° at least to the
specified set point temperature (in the Heat menu, on page 25) before it switches on.
Once the solar comes on it will start converging as it is heating. This ensures that it
will not continually be switching on and off.
Run: Set the temperature differential to stop heating from 2° to 5°. This setting sets
how close to the target set point temperature to switch off solar heat.

The key word is 'deviate'.  It isn't a fixed differential, but a differential based on the temp the heater last turned off.  The water has to be below the set point AND has to have fallen 'stop delta' degrees before it will turn on again.  

This leaves us with logic like:
  1.  For all cases in the heater to turn on, the solar temp > water temp.
  2.  When the heater turns off, note the water temp ("prevHeaterOffTemp").  This will only live in the application, not persisted.
  3.  If the water temp is below the set point, and the prevHeaterOffTemp - water temp > start temp delta, the heater should turn on.
  4.  The heater should run until it reaches the set point + stop delta(??)

I'm *not* as clear on the "Run" temp setting.  Does "how close to the target" mean +/- x degrees?  or does it mean run until we reach the set point + run (stop) delta?  Or (because it's Pentair), something else entirely.

Scenario 1:  First time running heater
Start temp delta: 3
Stop temp delta: 5
Water (Body) Temp: 75
Set Point: 83
prevHeaterOffTemp: _not set yet_
Result: Turn on heater and run until 88 (water temp > set point + stop delta); prevHeaterOffTemp = 88.

Scenario 2 (continued assumptions and result):  
Pool temp decreases to 85.
Do we turn on the solar here because the temp has fallen 3 degrees and we are below the 88 degree mark (set point + stop delta)?
Or do we wait until the pool falls below 83 degrees and then start heating to 88 again?

Scenario 3a (continued assumption and result):
Let's assume we do turn on the heater at 85 because it has fallen 3 degrees.  The solar heats up to 86, but then it gets cloudy, and the solar shuts off because the solar temp < water temp.  This will set the prevHeaterOffTemp to 86.  Now the water temp needs to drop to 83 for the solar to kick on.  (I'm thinking this scenario makes less sense than 3b below and we should just wait until body temp < set point AND prevHeaterOffTemp - body temp > start delta).

Scenario 3b (continued from Scenario 2):
Let's assume we don't turn on the heater again until the body temp < set point.  We would start heating the pool at <83 degrees and heat until 88(?).  Here the body temp < set point AND prevHeaterOffTemp - body temp > start delta.  Same result as Scenario 1.

Other considerations:
- Need to clear prevHeaterOffTemp flag if the body state changes (off/spa/etc).
- Do we need to clear the prevHeaterOffTemp flag at the end of the day?  End of the schedule?  I'm thinking no, but there could be a scenario where the pool temp doesn't change more than prevHeaterOffTemp for 24+ hours and it would still lock out the solar from coming on the next day.  

LMK your thoughts.

**Reply by CodeSlinger69** [reply]
While I appreciate your diligence in extracting as much logic as possible from the Pentair manual, I think we can do better. :) :) Even if I was running a Pentair solar controller, I'm not exactly sure what those words in that manual would mean in terms of actual operations.

Some requirements
Just restating what you said with a bit more detail to make sure we understand each other:
- We want to heat the water with solar when a) the water temp is below the setpoint, and b) there is enough solar energy to make a difference.
- When the water reaches its setpoint, solar should not reengage until the water temperature drops by the differential (like with home thermostats)
- If the solar heating is interrupted before reaching the setpoint (e.g. such as by clouds ), then point 1 can be reused subject to various latches/delays

Defining terms
This is where I think we are interpreting things differently, so I want to be careful here. Read below and see if my thoughts around start_delta and stop_delta make sense, as it appears from your post that you think about these settings in another way.
- "setpoint" = our target water temperature
- "water_temp" - from REM
- "solar_temp" - from REM
- "start_delta" allows for enough solar energy to make a difference. If our water_temp is 70 and the solar_temp is 71, engaging the solar heating is not going to meaningfully warm the water. So our solar_temp needs to be something greater than water_temp for solar heating to be effective. This is why I had mine set to 7 degrees. From what I can tell, this is how the logic in njsPC worked prior to the changes you made today.
- "stop_delta" was interpreted by me as when to turn off the solar due to not having enough solar energy, such as due to clouds. For example if solar is on, and solar_temp drops to something less than water_temp + stop_delta, then solar should be disengaged.
- "differential_latch" - a new internal variable to track when the setpoint has been reached. It is only set in that situation.
- "differential_temp" - a new user defined setting to determine when the water_temp has fallen enough that the differential_latch can be reset.

The logic in my head
- if solar==off AND water_temp < setpoint AND solar_temp > water_temp + start_delta AND differential_latch==false AND short_cycling_delay==false, then solar=on
- if solar==on AND solar_temp < water_temp + stop_delta, then solar=off and short_cycling_delay=true
- if solar==on AND water_temp >= setpoint, then solar=off and differential_latch=true
- if differential_latch==true AND water_temp <= setpoint - differential_temp, then differential_latch=false
- if the current schedule is changed manually OR a new schedule starts OR the setpoint is changed manually, then differential_latch=false and short_cycling_delay=false

A couple of notes
- Solar cooling at night and multiple bodies make this a bit more complicated, but this basic logic should be able to be extended to those situations. Let's get this right first of course.
- Regarding your observation that "The water has to be below the set point AND has to have fallen 'stop delta' degrees before it will turn on again". I would assert that anytime the water_temp is below the setpoint, we should engage the solar if there is enough solar energy (start_delta), as long as differential_latch==false and short_cycling_delay==false. 
- Your "prevHeaterOffTemp" variable: I feel that this duplicates what setpoint hopes to achieve. So what if the solar turned off due to clouds? As long as our setpoint has not been hit, we should still engage solar when appropriate (differential_latch and short_cycling_delay) regardless of the water_temp when it was turned off.

Does that make sense or are we still two trains passing in the night?

**Reply by tagyoureit** [reply]
The triggers for the solar work based off of the heatMode and setPoint.  The end user is never in control of the solar being on or off directly.  

I do think you are correct on the stop temp delta, and I do believe that's what the vague 'run' description is that I posted above.   

The latch will just be a temperature differential.  It should cover your differential_latch and differential_temp.  We don't need to be concerned about resetting the latch with a schedule or setpoint change because the system is constantly evaluating whether or not to turn on the heat.

I just pushed a new iteration.  Try it out and report back.

**Reply by CodeSlinger69** [reply]
Thanks for the update. I'll pull tomorrow and test once the sun rises.

> The triggers for the solar work based off of the heatMode and setPoint. The end user is never in control of the solar being on or off directly.

Clicking in the red circle below allows for immediate override of solar on/off and setpoint:

![Screenshot from 2024-04-15 20-25-55](https://github.com/tagyoureit/nodejs-poolController/assets/10177469/3a94ae92-1953-4b5b-929f-3ec17e63ad03)

**Reply by tagyoureit** [reply]
> Clicking in the red circle below allows for immediate override of solar on/off and setpoint:

That's not turning the solar on or off.  That sets the heat mode (and, yes, set point).  If you had a gas heater on the same body you would also see heater and solar preferred as other heat modes.  But the logic to actually turn on/off the heater is solely based on the set point.  When the heater turns on, you'll see the heat status update.  That's when the solar is actually on and the only way to force the solar on is to lower the set point below the body temp.

**Reply by CodeSlinger69** [reply]
```
// RSG 04.15.2024 - Updates to heater logic for start/stop deltas.  #925
// 1.  For all heating cases in order for the heater to turn on, the solar temp > water temp.
// 2.  If the water temp is below the set point, we want to turn on the heater
// 3.  But only if the prevHeaterOffTemp - water temp > start temp delta
// 4.  Also only if there is enough heat ('run') to make it worthwhile
// 5.  The heater should run until it reaches the set point + stop delta
// 6.  When the heater turns off, note the water temp ("prevHeaterOffTemp").  This will only live in the application, not persisted.

let hState: HeaterState = state.heaters.getItemById(heater.id);

if (state.temps.solar > body.temp // 1
    && body.temp < cfgBody.heatSetpoint // 2
    && (typeof hState.prevHeaterOffTemp === 'undefined' || ((hState.prevHeaterOffTemp - body.temp) > heater.startTempDelta)) // 3
    && (state.temps.solar - body.temp) > heater.stopTempDelta // 4
    && body.temp < cfgBody.heatSetpoint // 5
) {
```

As I was waiting for the sun to come out today, I was looking through this [code commit](https://github.com/tagyoureit/nodejs-poolController/commit/01be7a3d7877a8f783bf7f5244151bd094e55f25) in detail, and it made me reconsider the terms I've defined above. I was clearly not interpreting these ideas in line with your intent. My apologies for that, and with that in mind, I'd suggest the following changes to these terms. I'm just writing this down for myself so I can keep track. These values are in Settings/Heaters in dashPanel:

**"startTempDelta"** used when the solar is turned off for any reason, such as loss of solar energy (clouds), or reaching the setpoint, and prevents the solar from reengaging until the water temperature drops startTempDelta degrees from when the solar was disengaged.

-    Scenario 1: Setpoint reached. In this case this setting operates like a "temperature differential" in home furnace thermostats, and addresses my original point in this thread.
-    Scneario 2: Cloudy day. In this case, measured solar energy may fluctuate as clouds pass in front of the sun. The first cloud that passes by will cause the solar to disengage, and the current water temp will be stored in the internal prevHeaterOffTemp variable. Once the sun is out again, the solar will NOT reengage immediately, as the code logic first requires the water temperature to drop startTempDelta degrees. In my pool at least (in Arizona), water temperature never drops during the day, only overnight. From a practical sense then, the first cloud that passes will disengage solar heating for the rest of the day. dashPanel only allows this to be set to "4" as a minimum, so my workaround for now will be to edit the poolConfig.json and set this value to "0", so that my solar will reengage after the clouds pass and the sun comes back out. As a note, I think this scenario should be handled by the short cycling logic as clear and separate from the "temperature differential" logic, but I understand why it was done this way in order to unify these two different latching mechanisms.  However, since I am setting this to zero in config, I'm effectively commenting out this logic for now, and my original observation at the top of this thread will continue to occur (about the solar engaging for 10 seconds each minute after the setpoint is reached).

**"stopTempDelta"** is used to determine if enough solar energy is present to make a difference. The measured solar temp needs to be stopTempDelta degrees greater than the water temp for the solar heating to engage.

Thank you for continuing to work with me on this.

**Reply by CodeSlinger69** [reply]
Also, point 5 in the written logic:
5.  The heater should run until it reaches the set point + stop delta

Does not match the code:
`&& body.temp < cfgBody.heatSetpoint // 5`

**Reply by tagyoureit** [reply]
Thanks for pointing these out... I'll update the comments and I add tooltips to the UI because different platforms use different terms, but clarity will certainly help.

Here's the language from the Intellicenter manual...
![image](https://github.com/tagyoureit/nodejs-poolController/assets/1646840/48f43628-3a4c-4b90-b93f-dbd119df79c7)


> Solar Heating: Sta Effective solar heat depends on the solar and water temperature and the
start and stop differentials, When solar is available, then the system will rotate the 3-Way
Valve and turn on the solar booster pump if applicable or increase the RPM’s if an IntelliFlo®
Pump is present. This allows the water to flow to the solar collectors. If a pressure cleaner
pump is installed, the cleaner pump will turn off for 10 minutes every time solar turns ON.
The solar roof sensor must be 6° higher (based on Start/Stop Temp. Delta) than the water
temperature. The pump must be ON for a few minutes for operation. When the roof sensor
and the water temperature are within 3° of each other (based on Start/Stop Temp. Delta),
solar heat will switch off. The solar temperature start and stop differential settings are
factory at 6° and 3°. These differential settings are adjustable. Refer to your solar service
professional for the optimum setting. Changing the differentials could result in the system
not working properly.

(Yes, the manual has spelling issues - Sta Effective solar.  Oy.)

**Reply by CodeSlinger69** [reply]
I think we've lost our way on this discussion a bit. Let me roll things back. In [this post](https://github.com/tagyoureit/nodejs-poolController/discussions/925#discussioncomment-9123319) I defined stop/start deltas that are very much in line with the last quote from the manual. Extracting the relevant bits from Pentair's blurb:

> The solar roof sensor must be 6° higher than the water temperature. 
> When the roof sensor and the water temperature are within 3° of each other, solar heat will switch off. 

The good thing is that the original logic, before we started farting around in this thread, was in line with Pentair's wording:
```
if (body.temp < cfgBody.heatSetpoint &&
    state.temps.solar > body.temp + (hstate.isOn ? heater.stopTempDelta : heater.startTempDelta)) {
```

However, IMHO, the code as of [this commit](https://github.com/tagyoureit/nodejs-poolController/commit/01be7a3d7877a8f783bf7f5244151bd094e55f25) does not reflect that Pentair intent.

Now to bring this back, my original reason for this thread was to only handle the situation when the setpoint was hit. I had no complaints about how start/stop deltas were working. If clouds rolled in, the existing logic in concert with the short cycling logic worked fine.

For hitting a setpoint however, it should be similar to how home thermostats work: When the setpoint is hit, do not reengage the furnace until the temperature drops by the specified differential. For njsPC, I've leveraged the prevHeaterOffTemp class member that you've added, and built some logic to use it as a latching mechanism. In the below code, prevHeaterOffTemp is only set when the setpoint is reached, and is unset when the differential is met. That differential is hardcoded to 1 degree in the below code. It is also unset when clearPrevHeaterOffTemp() is called per your logic in NixieBoard.ts, which is fine.

Here is the code I am testing right now. I'll run it tomorrow once the sun gets hot again and see what happens. Or I might mess with manually posting REM temps tonight if I have energy. Most likely tomorrow though.

```
case 'solar':
    if (mode === 'solar' || mode === 'solarpref') {
        let hState: HeaterState = state.heaters.getItemById(heater.id);
        if (body.temp < cfgBody.heatSetpoint &&
            state.temps.solar > body.temp + (hstate.isOn ? heater.stopTempDelta : heater.startTempDelta) &&
            typeof hState.prevHeaterOffTemp === 'undefined'
        ) {
            isOn = true;
            body.heatStatus = sys.board.valueMaps.heatStatus.getValue('solar');
            isHeating = true;
        }

        /// snipped out the solar cooling code as I'm not worried about that until the heating code is locked in ///

        if (typeof hState.prevHeaterOffTemp === 'undefined') {
            if (body.temp >= cfgBody.heatSetpoint) hState.prevHeaterOffTemp = body.temp; 
        }
        else {
            if (hState.prevHeaterOffTemp - body.temp > 1) hState.prevHeaterOffTemp === undefined;
        }
    }
    break;
```

**Reply by rstrouse** [reply]
Realistically, I believe it hangs on to the solar temp setpoint + 1 so it shouldn't cycle on you.  This is how IntelliCenter actually handles this.  To drop a full degree in temperature over 20kgal of water is hard to do that is unless the temperature probe is after the heater.  Why are you fluctuating so much?

**Reply by tagyoureit** [reply]
Our prev code didn't take into account any start/stop temp deltas.  So the solar would cycle with 0.03 degrees difference (83 -> 82.97).  I'm not even sure the "acceptable range" through any of the OCP UI's allows <3 degrees of start differential so having a minimum safety net at 1 degree makes sense.

**Reply by CodeSlinger69** [reply]
Testing underway. Solar did not turn on automatically this morning due to the old assignment vs. evaluation bug. I can't tell you how many times I've done this in my life. 

This line:
`if (hState.prevHeaterOffTemp - body.temp > 1) hState.prevHeaterOffTemp === undefined;`

Should be:
`if (hState.prevHeaterOffTemp - body.temp > 1) hState.prevHeaterOffTemp = undefined;`

This needs to be fixed in the current committed code base as well. Controller/Boards/SystemBoard.ts, line 4164.

I'll watch carefully today and see what happens when the setpoint is hit.

**Reply by tagyoureit** [reply]
>  due to the old assignment vs. evaluation bug

![image](https://github.com/tagyoureit/nodejs-poolController/assets/1646840/98e98616-16f6-4ec2-ba90-3ac060fadf4d)

Pushed up the fix.  Thx.

**Comment by johnny2678** [comment]
Great to see the discussion on this.  I can't be certain, but guessing this has something to do with my pump frequently switching back and forth to/from the solar circuit.

Also, not to hijack - but these settings still don't work on ET4 setup.  This was just trying to change the start delta from 6 to 7. Results in a stacktrace:
![image](https://github.com/tagyoureit/nodejs-poolController/assets/8219743/018a24f5-3133-4c18-a6ad-c27b3b02ae37)

```
ApiError: Message aborted after 3 attempt(s): 165,32,16,33,162,3,7,0,194,2,100 
    at RS485Port.writeMessage (/home/pi/nodejs-poolController/controller/comms/Comms.ts:1010:31)
    at RS485Port.processWaitPacket (/home/pi/nodejs-poolController/controller/comms/Comms.ts:946:22)
    at RS485Port.processOutboundPackets (/home/pi/nodejs-poolController/controller/comms/Comms.ts:954:19)
    at RS485Port.processPackets (/home/pi/nodejs-poolController/controller/comms/Comms.ts:936:14)
    at Timeout. (/home/pi/nodejs-poolController/controller/comms/Comms.ts:989:52)
    at listOnTimeout (node:internal/timers:569:17)
    at processTimers (node:internal/timers:512:7)
```

Replay of the action: 
[replay (5).zip](https://github.com/tagyoureit/nodejs-poolController/files/15011184/replay.5.zip)

**Reply by tagyoureit** [reply]
@johnny2678 The problem is this line `out.setPayloadByte(2, ((obj.startTempDelta || 6) - 3 << 6) | ((obj.stopTempDelta || 3) - 2 << 1));` EasyTouchBoard.ts#2797.  The values startTempDelta=7 and stopTempDelta=3 result in an invalid value of 258.

Do you have another way of setting this and capturing the right packets from another device (Screenlogic or indoor control panel?).  I need to see what the correct logic is.  If you do, and can run through and capture all the start/stop temp deltas we should be able to figure it out.

**Reply by johnny2678** [reply]
> Do you have another way of setting this and capturing the right packets from another device (Screenlogic or indoor control panel?). 

Hey @tagyoureit, I think this is where we ended up when I reported it previously. Unfortunately, I don't have an indoor control panel and I don't see those settings exposed in ScreenLogic?

I suppose we could make a plea to the ET users in the broader njspc community...🤓😜

**Comment by CodeSlinger69** [comment]
Creating a new reply to my O.P. so it shows up on the bottom.

Status update: All good so far. Yesterday my water temp hit the setpoint of 83, the solar heating turned off, and the latching mechanism worked as expected to prevent the pump from cycling. Overnight my water temperature dropped to 79.8, which cleared the latch. When the solar energy was strong enough this morning, the solar heating turned back on and is currently warming the water as I type this.

The code I'm using is pasted below, and I would ask that this be commited to the repository. It is a very small change to the code that existed prior to this thread in that it only adds the latching (temperature differential) logic. The code that handled start delta and stop delta was working just fine before this thread, so I suggest we keep that original logic.

A couple of additional thoughts:

- I did not add the latch for nocturnal cooling, and I don't think its needed. I'm happy to debate this.
- I might suggest that the temperature differential that is currently hardcoded as "1" in the code be exposed as a setting next to start and stop delta. Many home termostats have this as a programmable setting.
- Does this kind of latching logic need to be added for gas heaters/heat pumps? I think so for the same reason as solar - this is just standard practice with home furnace thermostats, so why not with pool furnaces?

Thanks again for the support and discussion on this.
```
case 'solar':
    if (mode === 'solar' || mode === 'solarpref') {
        let hState: HeaterState = state.heaters.getItemById(heater.id);

        // solar heating
        if (body.temp < cfgBody.heatSetpoint &&
            state.temps.solar > body.temp + (hstate.isOn ? heater.stopTempDelta : heater.startTempDelta) &&
            typeof hState.prevHeaterOffTemp === 'undefined'
        ) {
            isOn = true;
            body.heatStatus = sys.board.valueMaps.heatStatus.getValue('solar');
            isHeating = true;
        }

        // nocturnal cooling
        else if (heater.coolingEnabled &&
             body.temp > cfgBody.coolSetpoint && 
             state.temps.solar < body.temp - (hstate.isOn ? heater.stopTempDelta : heater.startTempDelta) &&
             state.heliotrope.isNight
        ) {
            isOn = true;
            body.heatStatus = sys.board.valueMaps.heatStatus.getValue('cooling');
            isHeating = true;
            isCooling = true;
        }

        // solar heating temperature diffential latch
        if (typeof hState.prevHeaterOffTemp === 'undefined') {
            if (body.temp >= cfgBody.heatSetpoint) hState.prevHeaterOffTemp = body.temp;
        }
        else {
            if (hState.prevHeaterOffTemp - body.temp > 1) hState.prevHeaterOffTemp = undefined;
        }
    }
    break;
```

**Reply by tagyoureit** [reply]
I asked ChatGPT to help here... 
The first approach is what is in the repo, 2nd is what you have above.  Read below and then try to change my mind.  :)  https://chat.openai.com/share/d2b62b68-6238-45da-9ffa-8de004ff5e9d  (I removed the nocturnal cooling to keep the conversation more focused.)

------
User
compare the two different approaches.  and ask me any clarifying questions that you may need to help.

ChatGPT
Certainly! Let's compare the two approaches:

1. Approach 1:
- Activation Logic:
  - Considers multiple conditions for activation, including solar temperature, water temperature, temperature differentials, and previous heater off temperature.
  - Activates the heater if all conditions are met, ensuring efficient heating.
- Temperature Differential Handling:
  - Employs separate start and stop temperature differentials to control when the heating starts and stops.
  - Utilizes both start and stop temperature deltas to determine heating behavior.
- Previous Heater Off Temperature Management:
  - Tracks the previous heater off temperature to avoid frequent cycling of the heater.
  - Updates the previous heater off temperature when the heater turns off.
2. Approach 2:
- Activation Logic:
  - Simplifies the activation logic by considering fewer conditions, mainly focusing on the temperature differentials and the presence of a previous heater off temperature.
  - Activates the heater based on a temperature differential between the water and solar temperatures and the absence of a previous heater off temperature.
- Temperature Differential Handling:
  - Adjusts the temperature differential based on whether the heater is already on or off, potentially providing more flexibility in heating control.
- Previous Heater Off Temperature Management:
  - Sets the previous heater off temperature when the water temperature is above the set point, simplifying the management of this parameter.
  - Clears the previous heater off temperature if the temperature differential exceeds a certain threshold, possibly to prevent unnecessary delays in heater activation.

Clarifying Questions:

1. Are there any specific performance metrics or goals (e.g., energy efficiency, temperature stability) that the system aims to achieve?
2. How critical is it to avoid frequent cycling of the heater?
3. Is there any specific reason for the differences in temperature differential handling between the two approaches?
Understanding these aspects will help provide a more nuanced comparison and determine which approach better suits the requirements of the system.

User
Answers:
1.  The only goal is to heat the pool to the desired temperature
2.  It is of utmost criticality that frequent cycling is avoided.  
3.  The differences are because two developers are each trying to make their case for the best approach.

ChatGPT
Thank you for the clarification. Given the goal of heating the pool to the desired temperature while minimizing frequent cycling of the heater, let's analyze how each approach aligns with these requirements:

1. Approach 1:
- Advantages:
  - Rigorous activation logic with multiple conditions may result in more precise control over heater activation, potentially leading to better temperature regulation.
  - Tracking the previous heater off temperature helps prevent frequent cycling by ensuring that the heater remains off until the temperature conditions are suitable.
- Considerations:
  - The complex activation logic may introduce more opportunities for bugs or unintended behavior if not thoroughly tested.
  - The management of start and stop temperature differentials adds flexibility but also complexity to the system.
2. Approach 2:
- Advantages:
  - Simplified activation logic may reduce the risk of bugs and make the code easier to understand and maintain.
  - Adjusting the temperature differential based on the current state of the heater could provide more adaptive control, potentially leading to smoother operation.
- Considerations:
  - Simplifying the logic may sacrifice some precision in heater activation, which could affect temperature regulation.
  - Clearing the previous heater off temperature based on a threshold may introduce the risk of premature activation if the threshold is not carefully chosen.

Given the criticality of avoiding frequent cycling, Approach 1 seems more robust in ensuring that the heater remains off until the temperature conditions are suitable. However, Approach 2 offers simplicity and potentially more adaptive control.

To determine the best approach, it might be beneficial to conduct thorough testing, including simulations or real-world trials, to evaluate each approach's performance in achieving the desired temperature while minimizing frequent cycling. Additionally, considering factors such as code maintainability and ease of understanding could also inform the decision.

Do you need further analysis or assistance in evaluating these approaches?

User
which one would you use?

ChatGPT
Given the criticality of avoiding frequent cycling while efficiently heating the pool to the desired temperature, I would lean towards Approach 1. Here's why:

1. Robust Activation Logic: Approach 1 employs a comprehensive set of conditions to determine when to activate the heater, considering factors such as temperature differentials and the previous heater off temperature. This robust logic helps ensure that the heater remains off until the temperature conditions are suitable, minimizing the risk of frequent cycling.
2. Prioritizing Temperature Stability: The tracking of the previous heater off temperature in Approach 1 indicates a deliberate effort to prioritize temperature stability and avoid unnecessary heater cycling. By waiting until the temperature conditions are sufficiently favorable, this approach aims to maintain a more stable pool temperature.
3. Potential for Precise Control: While Approach 2 offers simplicity, Approach 1's more detailed activation logic may provide more precise control over heater activation, potentially resulting in better temperature regulation and energy efficiency.
4. Maintainability and Extensibility: Despite its complexity, Approach 1's logic is well-documented and structured, making it easier to understand and maintain over time. Additionally, the use of separate start and stop temperature differentials adds flexibility and extensibility to the system, allowing for fine-tuning of heating parameters as needed.

Overall, while Approach 1 may be more complex than Approach 2, its emphasis on temperature stability and precise control aligns better with the goal of avoiding frequent cycling while efficiently heating the pool. However, thorough testing and validation in real-world conditions would be essential to confirm its effectiveness in meeting the system requirements.

Would you like to explore any other aspects or considerations before making a decision?

**Reply by CodeSlinger69** [reply]
Can't argue with ChatGPT! If you are not interested in my logic, no hard feelings, I'll just maintain a fork for myself. My logic works for me, but I understand that you are building something for a community that is larger than just me, so additional considerations are required.

**Comment by CodeSlinger69** [comment]
As my final update on this, I'd like to acknowledge how well the solar logic worked yesterday. We had "Simpsons" clouds all day (big fluffy white things intermixed with blue skies), which allowed me to observe the start/stop delta functionality. As a reminder, there were two main scenarios discussed in this thread:

1. Temperature differential latching on hitting setpoint - as previously reported, this works great with the addition of the new latching code.
2. Drop in solar energy, such as by cloud passing in front of sun - subject of this post.

I am running the original code prior to this thread; the relevant line looks like this:
`state.temps.solar > body.temp + (hstate.isOn ? heater.stopTempDelta : heater.startTempDelta) &&`

My start delta is 7 and my stop delta is 4. Yesterday the setpoint was never reached, so the latching mechanism was never engaged. The solar was able to cycle on and off as the clouds passed in front of the sun. IMHO, the line of code above perfectly captures the intent as described in Pentair's documentation, and I can confirm it works (in my situation) to cycle solar on and off as the available solar energy fluctuates.

Thanks again for the help in understanding this logic and helping me tune the code to my needs.
